### PR TITLE
unit tests added for the util.go file in pkg/util directory

### PR DESCRIPTION
**Pull request by vjayaramrh**
**- What this PR does and why is it needed**

The PR adds unit tests (code  refactor is not the scope for this PR) for most of the functions defined in the following files:
- pkg/util/node_annotations.go [Dropped per review comment that tests are duplicate]
- pkg/util/pod_annotation.go[Dropped per review comment that tests are duplicate]
-pkg/util/subnet_annotations.go[Dropped per review comment that tests are duplicate]
-pkg/util/util.go

**- Special notes for reviewers**

This PR is rebased on top of https://github.com/ovn-org/ovn-kubernetes/pull/1371. The files relevant to this particular PR that need review are in the commit https://github.com/ovn-org/ovn-kubernetes/pull/1382/commits/702946ed8f1a99bf98f1670572f8cec7aa75c083?file-filters%5B%5D=
- pkg/util/node_annotations_unit_test.go [Dropped per review comment that tests are duplicate]
- pkg/util/pod_annotation_unit_test.go [Dropped per review comment that tests are duplicate]
- pkg/util/subnet_annotations_unit_test.go[Dropped per review comment that tests are duplicate]
- pkg/util/util_unit_test.go

**- How to verify it**

Pull this PR into your workspace
Navigate to go-controller/pkg/util directory
execute below commands
`go test -coverprofile cp.out`
`go tool cover -html=cp.out`
select files  node_annotations.go, pod_annotation.go, util.go, subnet_annotations.go from drop down to verify coverage

**- Description for the changelog**

### CODE REVIEW

**Comment by danwinship** [comment]
> The PR adds unit tests (code refactor is not the scope for this PR) for most of the functions defined in the following files:

But there already were unit tests for most of those files: `node_annotations_test.go`, `pod_annotation_test.go`, `subnet_annotations_test.go`. This is mostly duplicating those tests...

**Comment by vjayaramrh** [comment]
> > The PR adds unit tests (code refactor is not the scope for this PR) for most of the functions defined in the following files:
> 
> But there already were unit tests for most of those files: `node_annotations_test.go`, `pod_annotation_test.go`, `subnet_annotations_test.go`. This is mostly duplicating those tests...

@danwinship Good point, at the time(about 15days back) I opened the PR to write UT, the coverage numbers for node_annotations.go, pod_annotations.go and subnet_annotations_go at 77.5%, 79.3% and 47.9% respectively (refer below)
Screen Shot 2020-06-18 at 9 38 23 AM
I saw an opportunity to increase coverage with this PR to 96.6%, 93.5% and 83.1% respectively for those files as seen below using the testify assertions.
Screen Shot 2020-06-18 at 9 38 13 AM

I shall update the PR to remove the UTs for node_annotations.go, pod_annotations.go and subnet_annotations.go. WIll retain only the util_unit_test.go file.

**Comment by danwinship** [comment]
well, I mean you were likely testing cases that the original tests were missing... would be good to merge what you had done with what was already there.

**Comment by vjayaramrh** [comment]
> well, I mean you were likely testing cases that the original tests were missing... would be good to merge what you had done with what was already there.

@danwinship Thanks for clarifying, I shall push another PR with tests for those 3 files.

**Comment by trozet** [comment]
@danwinship current approach by @vjayaramrh is to add unit tests via testify, and remove the leftover duplicates in ginkgo later. That way it will be more obvious if he removes duplicate tests in ginkgo in a separate PR, we can see coveralls change 0% and its easy merge. Also reduces the risk of merge conflicts with these unit test patches, since he's only adding new unit test files for the most part.

**Comment by vjayaramrh** [comment]
> @vjayaramrh can you please squash your commits into one?

@trozet FYI I have squashed the commits into one

File: go-controller/pkg/util/util_unit_test.go [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+			inpSubnetStr: "192.168.1.12/24",
+			onRetArgsExecUtilsIface: []onCallReturnArgs{
+				{"RunCmd", []string{"*mocks.Cmd", "string", "[]string", "string", "string", "string"}, []interface{}{bytes.NewBuffer([]byte("")), bytes.NewBuffer([]byte("")), fmt.Errorf("RunOVNNbctl error")}},
+			},
+			onRetArgsKexecIface: []onCallReturnArgs{
+				{"Command", []string{"string", "string", "string", "string"}, []interface{}{mockCmd}},
+			},
+		},
+		{
+			desc:         "IPv6 CIDR, ovs-vsctl fails to list logical switch ports",
```
Comment (trozet) [inline go-controller/pkg/util/util_unit_test.go:145]: this says fails to list logical ports, but the errExpected is false?

Comment (trozet) [reply go-controller/pkg/util/util_unit_test.go:145]: so maybe the description should be "IPv6 CIRD, never excludes" and check for nil?

Comment (vjayaramrh) [reply go-controller/pkg/util/util_unit_test.go:145]: > so maybe the description should be "IPv6 CIDR, never excludes" and check for nil?

Looks like the intention of the code is to not to exclude any IPv6 addresses by checking for IPv6 address CIDR and skip further processing. Updating the test case name to be "IPv6 CIDR, never excludes" should be good I believe

File: go-controller/pkg/util/util_unit_test.go [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+import (
+	"bytes"
+	"fmt"
+	mock_k8s_io_utils_exec "github.com/ovn-org/ovn-kubernetes/go-controller/pkg/testing/mocks/k8s.io/utils/exec"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util/mocks"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"net"
+	"testing"
+)
```
Comment (girishmg) [inline go-controller/pkg/util/util_unit_test.go:12]: L4-11: can you please group the imports like we have done everywhere else in the project? standard library followed by github.com

File: go-controller/pkg/util/util_unit_test.go [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+
+			ret, e := GetNodeChassisID()
+			if tc.errExpected {
+				assert.Error(t, e)
+			} else {
+				assert.Greater(t, len(ret), 0)
+			}
+			mockExecRunner.AssertExpectations(t)
+			mockCmd.AssertExpectations(t)
+		})
```
Comment (girishmg) [inline go-controller/pkg/util/util_unit_test.go:111]: L109-110: Do we really need this calls? Since we are mocking the corresponding methods with expected function arguments  and the return values, they will always be 'True'? So, why assert? L104-107 is what is testing GetNodeChassiID(), right?

File: go-controller/pkg/util/util_unit_test.go [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+		inpSubnetStr            string
+		errExpected             bool
+		onRetArgsExecUtilsIface []onCallReturnArgs
+		onRetArgsKexecIface     []onCallReturnArgs
+	}{
+		{
+			desc:         "IPv4 CIDR, ovs-vsctl fails to list logical switch ports",
+			errExpected:  true,
+			inpNodeName:  "ovn-control-plane",
+			inpSubnetStr: "192.168.1.12/24",
```
Comment (girishmg) [inline go-controller/pkg/util/util_unit_test.go:136]: just so that it is accurate please use 192.168.1.0/24

Comment (girishmg) [reply go-controller/pkg/util/util_unit_test.go:136]: (and also please updated other places below as well)

File: go-controller/pkg/util/util_unit_test.go [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+	tests := []struct {
+		desc                    string
+		inpNodeName             string
+		inpSubnetStr            string
+		errExpected             bool
+		onRetArgsExecUtilsIface []onCallReturnArgs
+		onRetArgsKexecIface     []onCallReturnArgs
+	}{
+		{
+			desc:         "IPv4 CIDR, ovs-vsctl fails to list logical switch ports",
```
Comment (girishmg) [inline go-controller/pkg/util/util_unit_test.go:133]: s/ovs-vsctl/ovn-nbctl/

File: go-controller/pkg/util/util_unit_test.go [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+			},
+		},
+		{
+			desc:         "IPv6 CIDR, never excludes",
+			errExpected:  false,
+			inpNodeName:  "ovn-control-plane",
+			inpSubnetStr: "fd04:3e42:4a4e:3381::/64",
+		},
+		{
+			desc:         "IPv4 CIDR, haveMangementPort=true, ovs-nbctl command excludeIPs list empty",
```
Comment (girishmg) [inline go-controller/pkg/util/util_unit_test.go:151]: s/ovs-nbctl/ovn-nbctl/

File: go-controller/pkg/util/util_unit_test.go [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+					"RunCmd", []string{"*mocks.Cmd", "string", "[]string", "string", "string", "string", "string", "string", "string", "string", "string"}, []interface{}{bytes.NewBuffer([]byte("")), bytes.NewBuffer([]byte("")), nil},
+				},
+			},
+			onRetArgsKexecIface: []onCallReturnArgs{
+				{"Command", []string{"string", "string", "string", "string", "string", "string", "string", "string", "string"}, []interface{}{mockCmd}},
+				{"Command", []string{"string", "string", "string", "string", "string", "string", "string", "string", "string"}, []interface{}{mockCmd}},
+			},
+		},
+		{
+			desc:         "IPv4 CIDR, haveMangementPort=false, ovs-nbctl command with excludeIPs list populated, returns error ",
```
Comment (girishmg) [inline go-controller/pkg/util/util_unit_test.go:176]: s/ovs-nbctl/ovn-nbctl/

File: go-controller/pkg/util/util_unit_test.go (hunk 2 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/util_unit_test.go:12]: good catch, shall fix it.

File: go-controller/pkg/util/util_unit_test.go (hunk 3 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/util_unit_test.go:111]: The AssertExpectations ensures that whatever is specified with `On` and `Return`for the mocks are called as expected. I remember a case in one of the earlier PRs where I specified something incorrectly in the "On"/"Return" and the test would still pass. Having the AssertExpectations helped catch that and make the test more accurate/readable and avoid cut n paste errors for newer tests. For e.g; try modifying L188 from`nil` to `fmt.Errorf("")` (in a real life scenario, say a developer did copy paste and forgot to change from fmt.Errorf to nil ) and run the test case to verify that it fails. Next, if you comment out lines 238 and 239, the test will pass. Someone new that read the test case may end up getting confused and would like to avoid that if at all possible.

File: go-controller/pkg/util/util_unit_test.go (hunk 5 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/util_unit_test.go:133]: Sigh, copy and paste overlook.
Thanks for the catch, shall fix it.

File: go-controller/pkg/util/util_unit_test.go (hunk 4 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/util_unit_test.go:136]: ok, shall fix them

File: go-controller/pkg/util/util_unit_test.go (hunk 6 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/util_unit_test.go:151]: thanks again

File: go-controller/pkg/util/util_unit_test.go (hunk 7 above)
Comment (vjayaramrh) [reply go-controller/pkg/util/util_unit_test.go:176]: shall fix it

**Comment by vjayaramrh** [comment]
> @vjayaramrh the unit testing for this particular function UpdateNodeSwitchExcludeIPs() is not complete, right? It doesn't target the code for when config.HybridOverlay.Enabled is `true` and it also doesn't cover the following code as well, correct?
> 
> ```
> 	args := []string{"--", "--if-exists", "remove", "logical_switch", nodeName, "other-config", "exclude_ips"}
> 	if len(excludeIPs) > 0 {
> 		args = []string{"--", "--if-exists", "set", "logical_switch", nodeName, "other-config:exclude_ips=" + excludeIPs}
> 	}
> ```

@girishmg the config.HybridOverlay.Enabled path is not covered but the other part is covered as can be seen in the screen shot attached
Screen Shot 2020-06-27 at 1 19 25 PM

File: go-controller/pkg/util/util_unit_test.go (hunk 3 above)
Comment (girishmg) [reply go-controller/pkg/util/util_unit_test.go:112]: Thanks for the explaination.

**Comment by vjayaramrh** [comment]
> > @vjayaramrh the unit testing for this particular function UpdateNodeSwitchExcludeIPs() is not complete, right? It doesn't target the code for when config.HybridOverlay.Enabled is `true` and it also doesn't cover the following code as well, correct?
> > ```
> > 	args := []string{"--", "--if-exists", "remove", "logical_switch", nodeName, "other-config", "exclude_ips"}
> > 	if len(excludeIPs) > 0 {
> > 		args = []string{"--", "--if-exists", "set", "logical_switch", nodeName, "other-config:exclude_ips=" + excludeIPs}
> > 	}
> > ```
> 
> @girishmg the config.HybridOverlay.Enabled path is not covered but the other part is covered as can be seen in the screen shot attached
> Screen Shot 2020-06-27 at 1 19 25 PM

Added additional tests to up the coverage related to config.HybridOverlay.Enabled==true path
Screen Shot 2020-06-27 at 3 46 23 PM
